src/main.rs

```diff
@@ -0,0 +1,289 @@
+            Error::BadEpoch(epoch) =>
+                format!("Invalid epoch {epoch}"),
+            Error::Base64(e) =>
+                format!("invalid base64 encoding: {e}"),
+            Error::Oprf(e) =>
+                format!("PPOPRF error: {e}"),
+        };
+        let body = Json(ErrorResponse { message });
+        (StatusCode::BAD_REQUEST, body).into_response()
+    }
+}
+
+/// Process PPOPRF evaluation requests
+async fn randomness(
+    State(state): State<OPRFState>,
+    Json(request): Json<RandomnessRequest>,
+) -> Result<Json<RandomnessResponse>, Error> {
+    debug!("recv: {request:?}");
+    let state = state.read().map_err(|_| Error::LockFailure)?;
+    let epoch = request.epoch.unwrap_or(state.epoch);
+    if epoch != state.epoch {
+        return Err(Error::BadEpoch(epoch));
+    }
+    if request.points.len() > MAX_POINTS {
+        return Err(Error::TooManyPoints);
+    }
+    let prove = false;
+    let mut points = Vec::with_capacity(request.points.len());
+    for base64_point in request.points {
+        let input = BASE64.decode(base64_point)
+            .map_err(Error::Base64)?;
+        // FIXME: Point::from is fallible and needs to return a result.
+        // partial work-around: check correct length
+        if input.len() != ppoprf::COMPRESSED_POINT_LEN {
+            return Err(Error::BadPoint);
+        }
+        let point = ppoprf::Point::from(input.as_slice());
+        let evaluation = state.server.eval(&point, epoch, prove)
+            .map_err(Error::Oprf)?;
+        points.push(BASE64.encode(evaluation.output.as_bytes()));
+    }
+    let response = RandomnessResponse { points, epoch };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Process PPOPRF epoch and key requests
+async fn info(
+    State(state): State<OPRFState>
+) -> Result<Json<InfoResponse>, Error> {
+    debug!("recv: info reqeust");
+    let state = state.read().map_err(|_| Error::LockFailure)?;
+    let current_epoch = state.epoch;
+    // FIXME: return the end of the current epoch
+    let next_epoch_time = "unknown".to_owned();
+    let max_points = MAX_POINTS;
+    let public_key = state.server.get_public_key()
+        .serialize_to_bincode()
+        .map_err(Error::Oprf)?;
+    let public_key = BASE64.encode(public_key);
+    let response = InfoResponse {
+        current_epoch,
+        next_epoch_time,
+        max_points,
+        public_key,
+    };
+    debug!("send: {response:?}");
+    Ok(Json(response))
+}
+
+/// Advance to the next epoch on a timer
+async fn epoch_update_loop(state: OPRFState, epochs: EpochRange) {
+    let interval = std::time::Duration::from_secs(5);
+    info!("rotating epoch every {} seconds", interval.as_secs());
+
+    loop {
+        // Wait until the current epoch ends.
+        tokio::time::sleep(interval).await;
+
+        // Acquire exclusive access to the oprf state.
+        // Panics if this fails, since processing requests with an
```

Comment:
Panic means that the program exits, right? Should be fine considering that this will make Kubernetes restart the pod, which results in a sync with an existing enclave.